initrd: Opt-in bare bones systemd-based initrd

ElvishJerricco opened the pull request:
Third time's the charm.

# Motivation

This PR adds an option, `boot.initrd.systemd.enable`, that causes NixOS to generate an initrd that uses `systemd` for PID 1. This brings several benefits:

- Upstream units can be used to implement almost all of stage 1 logic, eliminating the need for the many lines of shell scripting NixOS has in its traditional stage 1.
- Early boot becomes parallelized.
- `systemd-ask-password` can be used to request credentials, providing a more streamlined way for early boot tasks to communicate with users.
- Configuration of stage 1 becomes declarative in exactly the same way that systemd in stage 2 is declarative, instead of adding imperative shell scripts in a variety of places.

## How this compares to my previous attempts

This version of a systemd-based initrd has three major improvements over the previous attempts.

1. It is an opt-in alternative to NixOS's traditional initrd, instead of a replacement.
2. It uses the same systemd option types that NixOS uses for stage 2 systemd configuration.
3. It will be implemented piecemeal, instead of in one large PR.
4. The produced initrd appears to be quite a bit smaller than the previous attempts for whatever reason, at just 9.1M (which is quite good compared to NixOS's tradtional initrd).

The third improvement is especially important. This PR only implements the *core functionality* needed to have a systemd-based initrd. There are many things that NixOS's traditional initrd supports that this does not yet. But with this framework in place, we can begin to add those features to the systemd-initrd one by one in separate, parallel PRs.

Notably, this PR does keep the `make-initrd-ng` concept from the previous one. You can check the README included with it in this PR for details, but the short version is that we don't copy full closures into initrd, and we don't patch things. Instead, we try to automate copying only the exactly correct things to their original paths to the best of our ability, and manually add what the automation misses in the `boot.initrd.systemd.objects` option. The benefits are outlined in the README.

This can be tested with `nix-build ./nixos -A vm`.

## What this approach will *not* do

is include a compatibility layer for options like `postDeviceCommands` and the like. Although this is possible in principle with special units that run the concatenated script with `Before=` and `After=` clauses on the appropriate targets, I think this is a bad idea for several reason.

1. It is antithetical to the purpose of using systemd. We want stage 1 to be parallelized, and declarative. These scripts are the opposite of both. It would slow down boot times and make the code for systemd-initrd less clean.
2. These scripts sometimes expect to be run in the user interactive console. Trying to accommodate this would be troublesome.
3. Maddness this way lies. I can only imagine the plethora of complexity and edge cases involved in trying to keep `extraUtils` and the `*Commands` options alive in this initrd.

## Next steps

Here are a number of things to do in future PRs, in very loose, decreasing order of how I view their importance.

- Testing. Currently this will only boot extremely simple setups. It'd be good to know *how* simple that is, and it will be important to test the other things we work on as we go to make sure they work as well.
- Support for the `script` style options in `boot.initrd.systemd.services`. Currently they do not copy the necessary scripts in.
- Proper udev rules configuration.
- Basic LUKS support.
- ZFS / btrfs / LVM / mdadm support.
- Plymouth support
- Networking / SSH support
- Advanced LUKS features (e.g. yubikeys).
- Move activation into stage 1 so that we can immediately switch-root to systemd. This is nice because systemd can serialize and deserialize some useful information if it knows it's switch-root'ing to systemd.

Most of these should be doable independently of each other, and likely need little more API than the core functionality provided by this PR.

---

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ElvishJerricco committed:
make-initrd-ng: init

ElvishJerricco committed:
nixos: systemd: Split unit types into separate module

ElvishJerricco committed:
nixos: systemd-lib: Make generateUnits general with default args

ElvishJerricco committed:
initrd: Optional systemd-based initrd

ElvishJerricco committed:
systemd-initrd: Partially fix qemu-vm

TODO: Auto formatting / partitioning / resizing

SuperSandro2000 commented on nixos/lib/utils.nix:
```diff
@@ -197,5 +197,6 @@ rec {
   systemdUtils = {
     lib = import ./systemd-lib.nix { inherit lib config pkgs; };
     unitOptions = import ./systemd-unit-options.nix { inherit lib systemdUtils; };
+    types = import ./systemd-types.nix { inherit lib systemdUtils; };
```
```suggestion
    lib = import ./systemd/lib.nix { inherit lib config pkgs; };
    unitOptions = import ./systemd/unit-options.nix { inherit lib systemdUtils; };
    types = import ./systemd/types.nix { inherit lib systemdUtils; };
```
Maybe? I am not sure. Just an idea.

ElvishJerricco committed:
systemd-initrd: Add PATH to everything

ElvishJerricco committed:
systemd-initrd: autoFormat and autoResize in initrd

ElvishJerricco committed:
systemd-initrd: Basic test case

mweinelt commented on nixos/tests/systemd-initrd.nix:
```diff
@@ -0,0 +1,11 @@
+import ./make-test-python.nix ({ pkgs, ... }: {
```
Should be added to all-tests.nix.

ElvishJerricco replied:
Done.

mweinelt commented on pkgs/build-support/kernel/make-initrd-ng.nix:
```diff
@@ -0,0 +1,79 @@
+  mkdir ./root
+  make-initrd-ng "$contentsPath" ./root
+  mkdir "$out"
+  (cd root && find * .[^.*] -exec touch -h -d '@1' '{}' +)
+  (cd root && find * .[^.*] -print0 | sort -z | cpio -o -H newc -R +0:+0 --reproducible --null | eval -- $compress >> "$out/initrd")
```
What do these find statements match? What is the benefit of using the wildcard `*` over `.` here? I think the regex could excludes `..`, which wouldn't be matched with `find . ` or `find root`.

ElvishJerricco replied:
@mweinelt A ton of this file was just copied from the old makeInitrd. In the case of the `find` pipeline, I figured there must have been *some* reason for it, so I did it the same way instead of what I thought would have been more obvious.

mweinelt replied:
cc @lheckemann @joachifm @abbradar

twhitehead replied:
The biggest difference I would see is that if you `find .` you

1. get `.`, and
2. your name are proceeded with a `./`.

From a bit of testing, the only difference it makes to the archive is whether it contains a `.` entry as the leading `./` are stripped from the names. I don't know if this matters at all, but I suspect not unless it somehow messes with ownership or permissions of the root directory when extracting.

I would also guess that the `.[^.*]` is a mistake (it matches two letter names where the first is `.` and the second is not `.` or `*`). Most likely the intent was `.[^.]*` to get all hidden names (i.e., all `.*` names with the exception of `..`).

Any, if you want the same behaviour, from a `find .` variant, it would be (`%P` excludes the specified starting point)

```
find . -printf '%P\0'
```

It will generate a warning from `cpio` about ignoring the empty line (which is `.` with the `.` suppressed). Given `cpio` strips the leading `./`, another option that just suppresses `.` would be

```
find . -path . -o -print0
```

ElvishJerricco committed:
systemd-initrd: Fix Environment= and PATH

ckiee commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,342 @@
+
+      targets.initrd.aliases = ["default.target"];
+      units =
+           mapAttrs' (n: v: nameValuePair "${n}.path"    (pathToUnit          n v)) cfg.paths
+        // mapAttrs' (n: v: nameValuePair "${n}.service" (initrdServiceToUnit n v)) cfg.services
+        // mapAttrs' (n: v: nameValuePair "${n}.slice"   (sliceToUnit         n v)) cfg.slices
+        // mapAttrs' (n: v: nameValuePair "${n}.socket"  (socketToUnit        n v)) cfg.sockets
+        // mapAttrs' (n: v: nameValuePair "${n}.target"  (targetToUnit        n v)) cfg.targets
+        // mapAttrs' (n: v: nameValuePair "${n}.timer"   (timerToUnit         n v)) cfg.timers
+        // listToAttrs (map
+                     (v: let n = escapeSystemdPath v.where;
+                         in nameValuePair "${n}.mount" (mountToUnit n v)) cfg.mounts)
+        // listToAttrs (map
+                     (v: let n = escapeSystemdPath v.where;
+                         in nameValuePair "${n}.automount" (automountToUnit n v)) cfg.automounts);
```
This feels pretty repetitive but it does keep the code simpler vs. a higher level approach. Not sure what to make of that.

ElvishJerricco replied:
Yea it's not the first time this is repeated in nixos. I suppose I could factor it out into a function. Would that be good?

ckiee replied:
Yeah, that'd be nice! The generic-unit-ification being in what is effectively the consumer of the systemd lib seems weird.

ckiee commented on nixos/tests/systemd-initrd.nix:
```diff
@@ -0,0 +1,11 @@
+  };
+
+  testScript = ''
+    machine.wait_for_unit("basic.target")
```
Maybe also test `autoResize`/`autoFormat`?

ElvishJerricco replied:
`autoFormat` is implicitly tested because the machines in a nixos test use it to format their root disks. But an `autoResize` test would be good as well.

ElvishJerricco replied:
I've added such a test

ckiee commented on nixos/lib/systemd-lib.nix:
```diff
@@ -396,6 +396,13 @@ in rec {
+          ${let env = def.environment;
+            in concatMapStrings (n:
+              let s = optionalString (env.${n} != null)
+                "Environment=${builtins.toJSON "${n}=${env.${n}}"}\n";
```
Cheeky.
```suggestion
                "Environment=${''"${n}=${env.${n}}"''}\n";
```

ElvishJerricco replied:
This is how it was done in the regular service version of this. For instance, what if your `env.${n}` has a `\n` character in it? With `toJSON` it'll get properly encoded, whereas that would corrupt the unit file.

ckiee replied:
Ah, didn't think of that.

ckiee reviewed (changes requested):
I started this review yesterday, there's a chance you already touched some of the things I reviewed.

Mostly code quality nitpicks, but more test{s,ing} for [this] core system component would be nice to see.

ElvishJerricco commented on pkgs/build-support/kernel/make-initrd-ng.nix:
```diff
@@ -0,0 +1,79 @@
+  # Get the basename of the actual compression program from the whole
+  # compression command, for the purpose of guessing the u-boot
+  # compression type and filename extension.
+  compressorName = fullCommand: builtins.elemAt (builtins.match "([^ ]*/)?([^ ]+).*" fullCommand) 1;
```
@ckiee This was part of the stuff I just copied from the old make-initrd.nix, but I believe the difference is with examples like `/foo/bin/gzip -9`. With `baseNameOf`, you get `gzip -9`, but with this you get `gzip`.

ckiee replied:
```
nix-repl> builtins.baseNameOf "/bin/hello bleh"
"hello bleh"
```
Seems so, sorry! :D

ElvishJerricco commented on pkgs/build-support/kernel/make-initrd-ng/src/main.rs:
```diff
@@ -0,0 +1,208 @@
+    let mut iter = p.components().peekable();
+    while let Some(comp) = iter.next() {
+        match comp {
+            Component::Prefix(_) => panic!("This tool is not meant for Windows"),
```
Is that the best way to ensure a crate isn't built for windows? Honestly it probably doesn't matter considering this is checked into nixpkgs and is very obviously intended only for linux.

flokli replied:
I think having it panic if someone compiles this for windows is fine :-)

ElvishJerricco commented on pkgs/build-support/kernel/make-initrd-ng/src/main.rs:
```diff
@@ -0,0 +1,208 @@
+) -> Result<(), Error> {
+    fs::copy(&source, target)?;
+
+    // If it's not a dynamically linked ELF file, we're done.
```
Sure, why not.

ElvishJerricco committed:
systemd-initrd: Test autoResize

ElvishJerricco commented:
#164016 has been merged, so I've rebased onto its last commit and removed the WIP label from this PR

dasJ commented on nixos/modules/system/boot/stage-1.nix:
```diff
@@ -706,8 +706,10 @@ in
-      { inherit bootStage1 initialRamdisk initialRamdiskSecretAppender extraUtils; };
+    system.build = mkMerge [
+      { inherit bootStage1 initialRamdiskSecretAppender extraUtils; }
+      (mkIf (!config.boot.initrd.systemd.enable) { inherit initialRamdisk; })
```
```suggestion
      (mkIf (!config.boot.initrd.systemd.enable) { inherit initialRamdisk; }) # generated in nixos/modules/system/boot/systemd/initrd.nix
```

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,366 @@
+
+  fstab = pkgs.writeText "fstab" (lib.concatMapStringsSep "\n"
+    ({ fsType, mountPoint, device, options, autoFormat, autoResize, ... }@fs: let
+        # TODO: We should use this in stage 2 as well
```
Maybe convert this to an issue because I find it unlinkely for somebody to look here when modifying stage 2

ElvishJerricco replied:
#165256 done

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,366 @@
+    allowMissing = false;
+  };
+
+  emergencyEnv = pkgs.buildEnv {
```
I find this to be highly misleading because it's not only used for emergency cases. Either rename it to something else or make it only be used for emergencies.

ElvishJerricco replied:
How about "utilities"?

dasJ replied:
`extraUtils` :P 
NixOS calls it `path` (as in `system.build.path`) but I don't like it because it's not clear whether this is `$PATH` or the path of the system.
What about `initrdBin`? Or just `bin` (since it's already in the initrd file)?

ElvishJerricco replied:
Sure, `initrdBin` seems fine.

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,366 @@
+      not yet supported by the intrd generated with this option.
+    '';
+
+    package = mkOption {
```
Either add `visible = false` to all of these or move them into a shallow submodule, like so:
```
options.boot.initrd.systemd = lib.mkOption {
  description = "…";
  visible = "shallow";
  type = submodule ({ ... }: {
    # …
  });
};
```

ElvishJerricco replied:
*Slightly* different than `mkPackageOption` because the default is `systemdMinimal` but the description should probably just say `systemd`.

dasJ replied:
That's not an issue:
```
lib.mkPackageOption pkgs "systemd" {
  default = "systemdMinimal";
};

ElvishJerricco replied:
Ah

Also, when I try to use this:

```
options.boot.initrd.systemd = lib.mkOption {
  description = "…";
  visible = "shallow";
  type = submodule ({ ... }: {
    # …
  });
};
```

I get infinite recursion

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,366 @@
+
+    objects = mkOption {
+      description = "List of objects to include in the initrd, and their symlinks";
+      example = ''
```
I think this should be a `literalExample`? Not sure

ckiee replied:
cc @pennae

pennae replied:
should be a `literalExpression` as it is written now (`literalExample` is deprecated)

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,366 @@
+        Blank for no password, ! for super user disabled.
+      '';
+      default = "!";
+    };
```
```suggestion
    };

```

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,366 @@
+      '';
+      default = "!";
+    };
+    emergencyPackages = mkOption {
```
Again, I find this misleading

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,366 @@
+      type = types.listOf types.str;
+      example = [ "systemd-backlight@.service" ];
+      description = ''
+        A list of units to suppress when generating system systemd configuration directory. This has
```
What does "suppress a unit" mean in this context? I find that unclear unless I look at the implementation

ElvishJerricco replied:
What would you suggest this say?

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,366 @@
+      description = "Definition of systemd socket units.";
+    };
+
+    timers = mkOption {
```
Do we even need these? I don't think it's incredibly important to drop these but it feels strange to me

ElvishJerricco replied:
It's the kind of thing that costs nothing to have and who knows what kinds of uses someone might come up with.

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,366 @@
+        { object = "${cfg.package.util-linux}/bin/sulogin"; }
+
+        {
+          object = builtins.toFile "system.conf" ''
```
Since you use that a lot - why not `pkgs.writeText`? That would allow files to have dependencies

ElvishJerricco replied:
Because these files don't have dependencies :)

dasJ commented on nixos/tests/systemd-initrd.nix:
```diff
@@ -0,0 +1,24 @@
+  machine = { pkgs, ... }: {
+    boot.initrd.systemd.enable = true;
+    fileSystems = lib.mkVMOverride {
+      "/".autoResize = true;
```
This is already the case, isn't it?

ElvishJerricco replied:
It is not. Currently, stage 1 doesn't have an implementation of auto resizing and formatting. The things we have that do this (like qemu-vm or sd-image) have their own special code for it. In fact, before this PR, setting `fileSystems."/".autoResize = true;` would result in errors because the code we have generating systemd units for resizing didn't handle "/" gracefully.

dasJ commented on nixos/tests/systemd-initrd.nix:
```diff
@@ -0,0 +1,24 @@
+import ./make-test-python.nix ({ lib, pkgs, ... }: {
```
Maybe call the entire test `systemd-initrd-simple` or something like this so it's more consistent when adding more tests in the future

dasJ reviewed (commented):
I haven't really looked at the Rust code apart from my one suggestion which helped me in debugging yesterday.

The approach is very promising and apart from the first commit not really splittable anymore. Most things I comment here are so that we nail some details right at the beginning instead of people starting to work on different aspects of the initrd while the core of it is still in flux.

I'm currently discussing stuff with @tfc to make sure we can get better test coverage. While I don't think this test needs more, I'd love future PRs with more features having the ability to add more stage 1 tests.

About the switch to the real system - I don't know if this PR should already contain code to switch to the new systemd directly without the shell script in between. While this approach in this PR already works of course, it's another thing people may make assumptions about ("this worked in systemd stage-1 before!!"). As you know, I'm experimenting with this detail and will provide my findings when they come up.

ElvishJerricco commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,366 @@
+  };
+
+  config = mkIf (config.boot.initrd.enable && cfg.enable) {
+    system.build = { inherit initialRamdisk stage1Units; };
```
I just found it quite useful for debugging the /etc/systemd/system directory

ElvishJerricco committed:
Clarify suppressed units description

flokli commented on nixos/lib/systemd-lib.nix:
```diff
@@ -286,17 +291,11 @@ in rec {
-  serviceConfig = { name, config, ... }: {
+  mkServiceConfig = path: { name, config, ... }: {
     config = mkMerge
       [ { # Default path for systemd services.  Should be quite minimal.
```
I assume this comment should be moved to the `serviceConfig =…` assignment below, as the list of packages isn't here anymore.

flokli replied:
Pushed to fc91cdb5bcd909f32496c4b69347fb1bd0755a77.

flokli commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,379 @@
+          symlink = "/etc/systemd/system.conf";
+        }
+
+        # TODO: Not sure why this needs to be here for the recovery shell to work
+        { object = "${pkgs.glibc}/lib/libnss_files.so"; }
```
This NSS module looks up usernames/groups from `/etc/groups` / `/etc/passwd`.

flokli replied:
documented in 74bae067487b1292edf5e97962de2bccb014ca15.

flokli commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,379 @@
+            (lib.concatStringsSep "\n" config.boot.initrd.kernelModules);
+        }
+        {
+          object = builtins.toFile "passwd" "root:x:0:0:System Administrator:/root:/bin/bash";
+          symlink = "/etc/passwd";
+        }
+        {
+          object = builtins.toFile "shadow" "root:${config.boot.initrd.systemd.emergencyHashedPassword}:::::::";
+          symlink = "/etc/shadow";
+        }
```
We might want to extract `fakeNss` from `pkgs/build-support/docker/default.nix` (which should fix looking up group names, too), and use this for /etc/groups and /etc/passwd - and only set `/etc/shadow` here.

flokli replied:
done in 74bae067487b1292edf5e97962de2bccb014ca15.

flokli commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,379 @@
+      # a mount unit on its makefs unit causes it to be unmounted when
+      # we isolate for switch-root. Use a dummy package so that
+      # generateUnits will generate drop-ins instead of unit files.
+      packages = [(pkgs.runCommand "dummy" {} ''
+        mkdir -p $out/etc/systemd/system
+        touch $out/etc/systemd/system/systemd-{makefs,growfs}@.service
+      '')];
+      services."systemd-makefs@".unitConfig.IgnoreOnIsolate = true;
+      services."systemd-growfs@".unitConfig.IgnoreOnIsolate = true;
```
Odd. Don't other distros with systemd-in-initrd run into the same problem? What do they use? Maybe we can link from here.

ElvishJerricco replied:
@flokli I'm not sure that other distros actually use `systemd-makefs` or `systemd-growfs` in initrd. [For instance, the fedora guide for installing a fedora image on an sd card for raspberry pi](https://docs.fedoraproject.org/en-US/quick-docs/raspberry-pi/#installing-fedora-on-a-raspberry-pi-for-linux-users_rpi) tells you to resize the root fs manually before booting the sd image on the pi.

flokli replied:
Hmmh, so we run the fstab generator in initrd, right?

It generates a unit file like this (as per https://github.com/systemd/systemd/blob/main/src/shared/generator.c#L466-L486 and https://github.com/systemd/systemd/blob/main/src/shared/generator.c#L532-L543):
```
        fprintf(f,
                "# Automatically generated by %s\n\n"
                "[Unit]\n"
                "Description=Make Swap on %%f\n"
                "Documentation=man:systemd-mkswap@.service(8)\n"
                "DefaultDependencies=no\n"
                "BindsTo=%%i.device\n"
                "Conflicts=shutdown.target\n"
                "After=%%i.device\n"
                "Before=shutdown.target %s\n"
                "\n"
                "[Service]\n"
                "Type=oneshot\n"
                "RemainAfterExit=yes\n"
                "ExecStart="SYSTEMD_MAKEFS_PATH " swap %s\n"
                "TimeoutSec=0\n",
                program_invocation_short_name,
                where_unit,
                escaped);
```

> # We need IgnoreOnIsolate, otherwise the Requires dependency of
> # a mount unit on its makefs unit causes it to be unmounted when
> # we isolate for switch-root. …

I don't think it's a `Requires` on the mount unit:

```
# Automatically generated by systemd-fstab-generator

[Unit]
Documentation=man:fstab(5) man:systemd-fstab-generator(8)
SourcePath=/etc/fstab
Before=local-fs.target
After=blockdev@dev-mapper-root.target

[Mount]
Where=/
What=/dev/mapper/root
Type=btrfs
Options=…
```

Could it be because we don't use `systemctl switch-root`?:

```
       switch-root ROOT [INIT]
           Switches to a different root directory and executes a new system manager process below it. This is intended for usage in initial RAM disks ("initrd"), and
           will transition from the initrd's system manager process (a.k.a. "init" process) to the main system manager process which is loaded from the actual host
           volume. This call takes two arguments: the directory that is to become the new root directory, and the path to the new system manager binary below it to
           execute as PID 1. If the latter is omitted or the empty string, a systemd binary will automatically be searched for and used as init. If the system manager
           path is omitted, equal to the empty string or identical to the path to the systemd binary, the state of the initrd's system manager process is passed to the
           main system manager, which allows later introspection of the state of the services involved in the initrd boot phase.
```

flokli replied:
I looked at the vm test logs:

```
machine # [  OK  ] Finished Cleanup udev Database.
machine # [  OK  ] Reached target Switch Root.
machine #          Starting Switch Root...
machine # [    1.881830] systemd-journald[73]: Received SIGTERM from PID 1 (systemd).
machine # 
machine # <<< NixOS Stage 2 >>>
machine # 
machine # [    1.967187] EXT4-fs (vda): re-mounted. Opts: (null). Quota mode: none.
machine # [    1.970678] booting system configuration /nix/store/c2birzmvm77j79lh5njrz18z2nrjg50i-nixos-system-machine-22.05.git.82102fc37da
machine # running activation script...
machine # setting up /etc...
machine # starting systemd...
machine # [    4.579760] systemd[1]: Inserted module 'autofs4'
machine # [    4.603075] systemd[1]: systemd 249.7 running in system mode (+PAM +AUDIT -SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified)
```

It looks like `initrd-switch-root.service` fails. It'd run `systemctl --no-block switch-root /sysroot`, but we don't use /sysroot inside the initrd, but /mnt-root. We then for some reason still run stage 2, which does exec systemd, instead of using `systemctl switch-root` - which might explain the `IgnoreOnIsolate` hacks we need to add currently.

dasJ replied:
It's called `/sysroot` on the systemd-based stage-1 and `/mnt-root` on the bash one

ElvishJerricco replied:
@flokli If I force stage 1 into emergency mode and run `systemctl show sysroot.mount`, I see `Requires=dev-vda.device systemd-makefs@dev-vda.service system.slice`, meaning it does have a `Requires=` dependency on the makefs service. I remember digging through the fstab-generator code for quite a while before I figured out how it worked (the unit files themselves shown by `systemctl cat ...` don't make it clear either), but unfortunately I don't remember the cause I found for this `Requires=` dependency. Regardless, it is there, so `sysroot` gets unmounted without this `IgnoreOnIsolate` stuff.

ElvishJerricco replied:
Ah, something creates a `/run/systemd/generator/sysroot.mount.requires` directory with a symlink to the makefs unit.

flokli replied:
I guess it's fine to initially keep this, but we should add a comment, and maybe revisit this while addressing https://github.com/NixOS/nixpkgs/issues/165649.

ElvishJerricco replied:
There is a comment, and I'm not sure what this has to do with #165649?

flokli replied:
Checking why we have to set `IgnoreOnIsolate` here.

I assume if we stop switching to stage2, but directly to the other systemd (via `systemctl switch-root`) things could work.

ElvishJerricco replied:
No, that won't have any effect on this. The problem is just that stage 1 will do `systemctl --no-block isolate initrd-switch-root.target`, which goes on a killing spree to stop all units. It'll do this whether we have systemd-to-systemd handoff or not. Once the makefs unit gets stopped, because of the Requires= dependency, /sysroot will get unmounted. So we need `IgnoreOnIsolate` so that the makefs unit never gets stopped

flokli replied:
OK. I feel like this smells like something upstream could handle more gracefully, and we might want to have a talk with them about it at some time, but nothing that should block this ;-)

flokli commented on nixos/lib/systemd-lib.nix:
```diff
@@ -382,6 +391,22 @@ in rec {
         '';
     };
 
+  initrdServiceToUnit = name: def:
```
This only differs from the `serviceToUnit` definition above in not setting the `X-ReloadIfChanged/X-RestartIfChanged` fields, no? Is there any harm in having these fields around?

 - If there is, could the `initrdServiceToUnit` be a function moved to `nixos/modules/system/boot/systemd/initrd.nix`?
 - If there is not, could we just use `serviceToUnit` there?

flokli replied:
I pushed e3083decc4dea10e34ac08b705ea74808cefa524, which removes initrdServiceToUnit and replaces it just with serviceToUnit, which still seems to work. PTAL.

flokli committed:
fakeNss: move to toplevel

Make this reachable from pkgs.fakeNss. This is useful outside docker
contexts, too.

https://github.com/NixOS/nixpkgs/pull/164943#discussion_r833220769

flokli committed:
systemd-initrd: use pkgs.fakeNss, document why we need libnss_files.so

flokli committed:
systemd-initrd, systemd-lib: drop initrdServiceToUnit

We can just use serviceToUnit here.

flokli committed:
nixos/lib/systemd-lib.nix: move comment back down to packages

https://github.com/NixOS/nixpkgs/pull/164943#discussion_r833215903

dasJ commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,376 @@
+        # TODO: Limit this to the bare necessities
+        { object = "${cfg.package}/lib"; }
+
+        { object = "${cfg.package.util-linux}/bin/mount"; }
```
Coming to think of it, why are we not using busybox here? :eyes:

ElvishJerricco replied:
systemd has the path to util-linux's mount hard coded in its binary. Plus, busybox wouldn't make the initrd any smaller (I've tried via overriding systemd with `util-linux = busybox`)

ElvishJerricco commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,376 @@
+        }
+
+        { object = config.environment.etc.os-release.source; symlink = "/etc/initrd-release"; }
+        { object = config.environment.etc.os-release.source; symlink = "/etc/os-release"; }
```
@dasJ Without `os-release` we get `Welcome to Linux!` for the greeting instead of `Welcome to NixOS 22.05 (Quokka)!` Not a big deal, but it's kind of nice.

roberth commented on nixos/lib/systemd-lib.nix:
```diff
@@ -325,6 +324,17 @@ in rec {
       ];
   };
 
+  # Default path for systemd services. Should be quite minimal.
+  serviceConfig = mkServiceConfig [
+    pkgs.coreutils
+    pkgs.findutils
+    pkgs.gnugrep
+    pkgs.gnused
+    systemd
+  ];
```
```suggestion
  # Default path for systemd services. Should be quite minimal.
  bootedServiceConfig = {
    imports = [ serviceConfig ];
    path = mkAfter [
      pkgs.coreutils
      pkgs.findutils
      pkgs.gnugrep
      pkgs.gnused
      systemd
    ];
  };
```

roberth replied:
This must also be renamed elsewhere if you apply it as suggested.

roberth commented on nixos/lib/systemd-lib.nix:
```diff
@@ -286,17 +291,11 @@ in rec {
-            ];
-          environment.PATH = "${makeBinPath config.path}:${makeSearchPathOutput "bin" "sbin" config.path}";
+      [ {
+          path = mkAfter path;
```
```suggestion
```

roberth commented on nixos/lib/systemd-lib.nix:
```diff
@@ -286,17 +291,11 @@ in rec {
   };
 
-  serviceConfig = { name, config, ... }: {
+  mkServiceConfig = path: { name, config, ... }: {
```
```suggestion
  serviceConfig = { name, config, ... }: {
```

roberth reviewed (commented):
Let's use module composition instead of generated modules.
It keeps the types simple.

roberth commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,376 @@
+        [ { object = "''${systemd}/lib/systemd/systemd"; symlink = "/init"; } ]
+      '';
+      visible = false;
+      type = types.listOf (types.submodule {
```
Please use `attrsOf` instead of putting the key in an option. That way, overriding is possible.

`symlink` is a key, but its name suggests that an alternative method of inclusion might be possible, such as coping perhaps? However, that will still involve a path in the initrd, which is therefore also a functional key. This is only more reason to use `attrsOf` for the path, so you don't have to write custom logic to deal with that.

I suppose another method of inclusion is "just add it only to the store", which is generally _not_ a valid thing to do. Instead we ought to gather the store contents automatically. It's ok to provide an escape hatch, but that shouldn't influence the general design too much. It could be a separate option, just like we have one for the regular store (`system.extraDependencies`).

Perhaps this could be modeled after `environment.etc`? It doesn't have to implement all its features, because it doesn't need to and its implementation can/should be simpler. A familiar interface is always nice.

roberth commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -0,0 +1,376 @@
+    };
+
+    emergencyHashedPassword = mkOption {
+      type = types.str;
```
`coercedTo (enum [ null ]) (_: "!") str` would make a lot of sense conceptually, fwiw.

dasJ replied:
Is what I just pushed a viable alternative?

roberth replied:
Not just viable, but better than what I was philosophizing about :)

The lack of tagged unions makes me uneasy sometimes, but whatever.

:+1:

dasJ committed:
nixos/systemd-lib: Use module composition

dasJ committed:
nixos/systemd-initrd: Remove unit options that don't work

dasJ commented:
@roberth can you take another look? I added two commits, one that does the type composition as requested and another one that adds type composition to the unit types to get rid of options that are not used

dasJ committed:
nixos/systemd-initrd: Make emergency access more flexible

dasJ committed:
nixos/systemd-initrd: Redo object specifications

As requested by @roberth, we now have an option similar to
environment.etc. There's also extra store paths to copy and a way to
suppress store paths to make customizations possible.

We also link mount and umount to /bin to make recovery easier when
something fails

dasJ committed:
nixos/systemd-stage-1: Append (Initrd) to /etc/initrd-release

This is more in line with what dracut does (it appends "Initramfs") and
makes it clear where the boot is currently at when it hangs.

ElvishJerricco commented:
@dasJ Those changes look great!

dasJ commented:
If there are no more objections and pushes, I'll merge this in the next 24-24h

bobvanderlinden commented:
Awesome! This unlocks so much potential improvements. Thanks a lot for all the work of everyone! 🥳🥳🥳

ncfavier commented:
Tried this out on my laptop, it's almost usable except for two issues:

- https://github.com/NixOS/nixpkgs/issues/178340
- https://github.com/NixOS/nixpkgs/issues/178345

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/is-gpt-partition-automounting-possible-such-that-dev-disk-by-id-is-not-necessary/34790/4
